transformations/suspecting_paraphraser/transformation.py

```diff
@@ -0,0 +1,186 @@
+from typing import Tuple, List
+import nltk
+from pyinflect import getInflection
+import spacy
+import gender_guesser.detector as gender
+from enum import Enum
+import numpy as np
+
+
+from interfaces.QuestionAnswerOperation import QuestionAnswerOperation
+from tasks.TaskTypes import TaskType
+
+"""
+Base Class for implementing the different input transformations a generation should be robust against.
+"""
+
+
+class Pronouns:
+    HE = 0
+    SHE = 1
+    IT = 2
+    THEY = 3
+
+
+class SuspectingParaphraser(QuestionAnswerOperation):
+    tasks = [TaskType.QUESTION_ANSWERING, TaskType.QUESTION_GENERATION]
+
+    languages = ["en"]
+
+    def __init__(self, seed=0, max_output=1, pronoun_mod=0.9):
+        super().__init__(seed)
+        np.random.seed(seed)
+        nltk.download("punkt")
+
+        self.seed = seed
+        self.max_output = max_output
+
+        self.nlp = spacy.load("en_core_web_sm")
+        self.gender_detector = gender.Detector()
+        self.pronouns = ["he", "she", "it", "they"]
+        self.static_pronouns = ["i", "you", *self.pronouns]
+
+        self._special_endings = {
+            "may": "may {} not",
+            "might": "might {} not",
+            "shall": "shan't {}",
+            "will": "won't {}",
+        }
+
+        self.pronoun_mod = pronoun_mod
+        self._pronoun_alt = (1 - pronoun_mod) / (len(self.pronouns) - 1)
+
+    def _transform(self, question):
+        text = nltk.word_tokenize(question)
+        doc = self.nlp(question)
+        modal = str(doc[0]).lower()
+        token = doc[0]
+
+        verb_position = [i for i in range(len(doc)) if str(doc[i]) == token.head.text][
+            0
+        ]
+
+        rest_of_sentence = [i for i in text[verb_position:]]
+
+        if text[1].lower() == "n't":
+            text = text[1:]
+            doc = doc[1:]
+            verb_position = verb_position - 1
+
+        beginning = [text[1].capitalize()]
+        beginning.extend(text[2:verb_position])
+        sentence = nltk.tokenize.treebank.TreebankWordDetokenizer().detokenize(
+            beginning + rest_of_sentence
+        )
+
+        first_verb = doc[verb_position]
+        if modal == "did":
+            demodded = first_verb._.inflect("VBD")
+        else:
+            demodded = modal + " " + str(first_verb)
```

Comment:
Could you add comments here explaining what the VBD pos tag means, and what this if else does in general [I guess either use modal + raw form or just the inflected form, but might help readers understannd/extend this easily]

Reply:
Done!